src/rewards/poprewards_cache.cpp

```diff
@@ -0,0 +1,126 @@
+// Copyright (c) 2019-2020 Xenios SEZC
+// https://www.veriblock.org
+// Distributed under the MIT software license, see the accompanying
+// file LICENSE or http://www.opensource.org/licenses/mit-license.php.
+
+#include <veriblock/rewards/poprewards_cache.hpp>
+
+namespace altintegration {
+
+PopRewardsBigDecimal PopRewardsCache::calculateDifficulty(
+    const index_t& /* ignore */) const {
+  auto it = buffer.crbegin();
+  // remove last element - it is the current block that we calculate difficulty
+  // for
+  ++it;
+  PopRewardsBigDecimal sumscore = 0.0;
+  size_t count = 0;
+  uint64_t difficultyInterval =
+      altParams_->getRewardParams().difficultyAveragingInterval();
+  for (; it != buffer.crend(); ++it) {
+    if (count >= difficultyInterval) {
+      break;
+    }
+    count++;
+    sumscore += it->second;
+  }
+
+  auto difficulty = sumscore / difficultyInterval;
+
+  // Minimum difficulty
+  if (difficulty < 1.0) {
+    difficulty = 1.0;
+  }
+  return difficulty;
+}
+
+std::map<std::vector<uint8_t>, int64_t> PopRewardsCache::calculatePayouts(
```

Comment:
it would probably be easier to overload just 1 function, scoreFromEndorsements, as that's what we are caching and hide all cache handling code there

Reply:
Nah, we won't be able to detect reorgs this way. For example we calculate score for block 90 and save it to cache. Next time we try to reuse the cache but we had a reorg and now block 90 has different score.
Therefore we rely on calculatePayouts() call that gives us a signal about currently processed block.

Reply:
It is better to trigger cache invalidation explicitly rather than assume that calculatePayouts now(in the cache) means invalidateCache + calculatePayouts